src/backend/access/transam/varsup.c

```diff
@@ -514,6 +515,67 @@ ForceTransactionIdLimitUpdate(void)
 	return false;
 }
 
+Oid
+GetNewTempObjectId(void)
+{
+	Oid			result;
+	Oid			tempOidStart;
+	Oid			real_temp_oid_buffer_start;
+	static Oid 	nextTempOid = InvalidOid;
+
+	/* safety check, we should never get this far in a HS standby */
+	if (RecoveryInProgress())
+		elog(ERROR, "cannot assign OIDs during recovery");
+
+	if (!OidIsValid((Oid) temp_oid_buffer_start)) /* InvalidOid means it needs assigning */
+	{
+		/* First check to see if another connection has already picked a start, then update. */
+		LWLockAcquire(OidGenLock, LW_EXCLUSIVE);
+		if (OidIsValid(ShmemVariableCache->tempOidStart) && (ShmemVariableCache->tempOidBufferSize == temp_oid_buffer_size))
+		{
+			/* If ShmemVariableCache->tempOidStart is valid, it could also mean the buffer size GUC has been altered, and we need to pick a new start. */
+			temp_oid_buffer_start = ShmemVariableCache->tempOidStart + INT_MIN;
+			nextTempOid = ShmemVariableCache->tempOidStart;
+		}
+		else
+		{
+			/* We need to pick a new start for the buffer range. */
+			tempOidStart = ShmemVariableCache->nextOid;
```

Comment:
Does this logic need to handle the case `if (ShmemVariableCache->nextOid < ((Oid) FirstNormalObjectId))` similar to `GetNewObjectId()`. Consider the case when `ShmemVariableCache->nextOid == 0` then we will end up using oids below `FirstNormalObjectId`